### PR TITLE
#60 Write an api to fetch users by author

### PR DESCRIPTION
**Pull request by aashishpiitk**

### CODE REVIEW

File: users/serializers.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+from posts.serializers import PostSerializer
+
+class PostByAuthorSerializer(serializers.ModelSerializer):
+    post_set=PostSerializer(many=True)
+    class Meta:
+        model=User
+        fields=(
+            'post_set',
+            'username'
+        )
```
Comment (namangup) [inline users/serializers.py]: New line

File: users/serializers.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+from rest_framework import serializers
+from .models import User
+from posts.serializers import PostSerializer
+
+class PostByAuthorSerializer(serializers.ModelSerializer):
```
Comment (namangup) [inline users/serializers.py]: PostByUser would be a more appropriate name ?

File: users/views.py [hunk 3]
```diff
@@ -241,3 +242,14 @@ def ResetPassword(request,token):
             return Response(response,status=status.HTTP_401_UNAUTHORIZED)
     else:
         return HttpResponse("Invalid Request",status=400) 
+
+
```
Comment (namangup) [inline users/views.py]: Remove extra whitespace.

File: users/views.py [hunk 4]
```diff
@@ -241,3 +242,14 @@ def ResetPassword(request,token):
...
+class PostsByAuthorView(APIView):
+    def get(self,request,pk):
+        try:
+            user = User.objects.get(pk=pk)
+            serializer = PostByAuthorSerializer(user)
+            return Response(serializer.data)
+        except:
+            return Response(status=status.HTTP_404_NOT_FOUND)
```
Comment (namangup) [inline users/views.py]: Add a new line here.

**Review by namangup — CHANGES_REQUESTED** [review]

File: users/serializers.py (hunk 2 above)
Comment (aashishpiitk) [reply users/serializers.py]: Done

File: users/views.py (hunk 4 above)
Comment (aashishpiitk) [reply users/views.py]: Done

File: comments/urls.py [hunk 5]
```diff
@@ -0,0 +1,7 @@
+from django.urls import path
+from .views import *
+
+urlpatterns = [
+    path('create/', CreateComment.as_view(), name="follow_stream"),
+    path('delete/', DeleteComment.as_view(), name="posts_by _stream"),
```
Comment (d1vyanshu) [inline comments/urls.py]: Why are the names follow_stream and posts_by_stream?

File: comments/views.py [hunk 6]
```diff
@@ -1,3 +1,84 @@
...
+                return Response(status=status.HTTP_404_NOT_FOUND)
+            try:
+                parent=request.data['parent_id']
+                parent_comment=Comment.objects.get(pk=parent)
+                comment=Comment.objects.filter(user=user,post=post,parent=parent_comment)
+                if len(comment) is not 0:
+                    response={
+                        "message":"comment from this user already exists"
+                    }
+                    return Response(response,status=status.HTTP_400_BAD_REQUEST)
```
Comment (d1vyanshu) [inline comments/views.py]: I think the user will not be able to reply more than once on a same comment. This should not be the case and users should be able to comment as many times as they wants.

**Comment by d1vyanshu** [comment]
You have to write the details of the api in readme file. Please ensure you give new line in all files you create.

File: comments/urls.py (hunk 5 above)
Comment (aashishpiitk) [reply comments/urls.py]: corrected

File: comments/views.py (hunk 6 above)
Comment (aashishpiitk) [reply comments/views.py]: ok

File: README.md [hunk 7]
```diff
@@ -49,3 +49,27 @@ To display posts corresponding to a stream
 url : /streams/<int:pk>/posts/
 method : GET
 ```
+#### Fetch Posts by User
+To display posts corresponding to a user
+```
+url:/users/<int:pk>/posts
+method: GET
+```
```
Comment (namangup) [inline README.md]: Move this under the user heading.

File: .gitignore [hunk 8]
```diff
@@ -1,3 +1,4 @@
 __pycache__/
 /**/__pycache__
 .vscode
+**/migrations/*
```
Comment (namangup) [inline .gitignore]: New line again.

File: README.md [hunk 9]
```diff
@@ -49,3 +49,27 @@ To display posts corresponding to a stream
...
+#### Bookmark/Unbookmark
+To bookmark a post or to unbookmark already existing bookmark
+```
+url:/bookmarks/create/
+method: POST
+```
```
Comment (namangup) [inline README.md]: Under a separate Bookmarks heading.
Add parameters.
Also I think there should be separate views to create / remove a bookmark . @d1vyanshu

File: README.md [hunk 10]
```diff
@@ -49,3 +49,27 @@ To display posts corresponding to a stream
...
+#### Comment
+To comment on post or sub-comment on a comment
```
Comment (namangup) [inline README.md]: Make this 
`
### Comment
#### Create Comment
`

File: README.md [hunk 11]
```diff
@@ -49,3 +49,27 @@ To display posts corresponding to a stream
...
+```
+url:/comments/create/
+method: POST
```
Comment (namangup) [inline README.md:78]: Add a description of request parameters.

File: README.md [hunk 12]
```diff
@@ -49,3 +49,27 @@ To display posts corresponding to a stream
...
+```
+#### Delete Comment
+To delete a comment on a post(all sub-comments will be deleted)/delete sub-comments(all of its sub-comments will be deleted).Recursive deletion will be followed
+```
+url:/comments/delete
+method: DELETE
```
Comment (namangup) [inline README.md:88]: Parameters

File: bookmark/views.py [hunk 13]
```diff
@@ -1,3 +1,41 @@
...
+from rest_framework.response import Response
+from rest_framework.parsers import JSONParser
+from django.shortcuts import render, redirect
+from django.http import HttpResponse,JsonResponse
+from django.views.decorators.csrf import csrf_exempt
+import bcrypt
+from posts.models import Post
+from bookmark.models import Bookmark
+from rest_framework.decorators import api_view, renderer_classes
+from rest_framework.views import APIView
```
Comment (namangup) [inline bookmark/views.py:6]: Remove unnecessary imports.

File: campusdiscussbackend/settings.py [hunk 14]
```diff
@@ -106,7 +106,7 @@ def get_secret(setting, secrets=secrets):
 DATABASES = {
     "default": {
         "ENGINE": "django.db.backends.postgresql_psycopg2",
-        "NAME": "postgres",
+        "NAME": "campus7",
```
Comment (namangup) [inline campusdiscussbackend/settings.py]: Why is this required ?

File: comments/views.py [hunk 15]
```diff
@@ -1,3 +1,72 @@
...
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from rest_framework.parsers import JSONParser
+from rest_framework.decorators import api_view, renderer_classes
+from rest_framework.renderers import JSONRenderer, TemplateHTMLRenderer
+from django.http import HttpResponse, JsonResponse
+from django.views.decorators.csrf import csrf_exempt
+from users.models import User
+from posts.models import Post
+from .models import Comment
```
Comment (namangup) [inline comments/views.py:7]: Again, remove unnecessary imports.

File: comments/views.py [hunk 16]
```diff
@@ -1,3 +1,72 @@
...
+    def post(self,request):
+        try:
+            user=IsLoggedIn(request)
+            if user is None:
+                return Response(status=status.HTTP_401_UNAUTHORIZED)
+            post_id=request.data['pk']
+            content=request.data['content']
+            post=Post.objects.get(pk=post_id)
+            if post is None:
+                return Response(status=status.HTTP_404_NOT_FOUND)
```
Comment (namangup) [inline comments/views.py:27]: Same issue.

File: comments/views.py [hunk 17]
```diff
@@ -1,3 +1,72 @@
...
+                parent_comment=Comment.objects.get(pk=parent)
+                Comment.objects.create(user=user,content=content,post=post,parent=parent_comment)
+                return Response(status=status.HTTP_200_OK)
+            except: 
+                Comment.objects.create(user=user,content=content,post=post)
+                return Response(status=status.HTTP_200_OK)
+        except:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+
+
```
Comment (namangup) [inline comments/views.py:33]: 1 Line is enough.

File: posts/views.py [hunk 18]
```diff
@@ -55,3 +56,14 @@ def delete(self, request):
                 return Response(status = status.HTTP_400_BAD_REQUEST)
         else:
             return Response(status=status.HTTP_401_UNAUTHORIZED)
+
+
+class PostDetailView(APIView):
+
```
Comment (namangup) [inline posts/views.py:62]: Add details to readme.

File: posts/views.py [hunk 19]
```diff
@@ -55,3 +56,14 @@ def delete(self, request):
...
+    def get(self,request,pk):
+        post=Post.objects.get(pk=pk)
+        if post is None:
```
Comment (namangup) [inline posts/views.py:65]: Put this in a try block, remove the if condition.

File: README.md (hunk 9 above)
Comment (d1vyanshu) [reply README.md]: Cannot think of any disadvantage as of now of keeping both under same view except of course it decreases the readability of code.

**Comment by d1vyanshu** [comment]
@aashishpiitk You need to keep api under different heading in readme. For example api concerned with some action on user model should be kept under user heading and so on. Also in PUT, POST requests you are supposed to mention the parameters you are passing along with request. If the url is supposed to be /delete/id then you need to mention the complete url so that we know the correct syntax to hit api. We don't want to go and lookup in code which wastes time. In some views there are reduntant import lines, you should keep the code clean and remove unnecessary redundancy.

File: bookmark/views.py [hunk 20]
```diff
@@ -1,3 +1,32 @@
...
+
+class CreateAndDeleteBookmark(APIView):
+    
+    def post(self,request):
+        try:
+            pk=request.data['pk']
+            post=Post.objects.get(pk=pk)
+            if post is None:
+                return Response(status=status.HTTP_404_NOT_FOUND)
```
Comment (namangup) [inline bookmark/views.py]: This if condition won't do anything. Read [this](https://docs.djangoproject.com/en/3.0/topics/db/queries/#retrieving-a-single-object-with-get). Django doesn't return None if object is not found.

File: comments/views.py [hunk 21]
```diff
@@ -1,3 +1,67 @@
...
+    def post(self,request):
+        try:
+            user=IsLoggedIn(request)
+            if user is None:
+                return Response(status=status.HTTP_401_UNAUTHORIZED)
+            post_id=request.data['pk']
+            content=request.data['content']
+            post=Post.objects.get(pk=post_id)
+            if post is None:
+                return Response(status=status.HTTP_404_NOT_FOUND)
```
Comment (namangup) [inline comments/views.py]: Same issue here.

**Review by namangup — COMMENTED** [review]
Follow PEP8 standards in your code.
Use spaces around operators and after commas, but not directly inside bracketing constructs.
Read [this](https://www.geeksforgeeks.org/pep-8-coding-style-guide-python/)

**Comment by namangup** [comment]
@aashishpiitk make sure that creating a new comment changes the last_modified field in the posts model.

File: bookmark/views.py (hunk 20 above)
Comment (aashishpiitk) [reply bookmark/views.py]: Done

File: comments/views.py (hunk 21 above)
Comment (aashishpiitk) [reply comments/views.py]: Done

**Comment by aashishpiitk** [comment]
> @aashishpiitk make sure that creating a new comment changes the last_modified field in the posts model.

did you confirm with anyone?
@namangup

**Comment by namangup** [comment]
> did you confirm with anyone?
> @namangup

I guess that is the only use of that field in the model ? 
@d1vyanshu

**Comment by namangup** [comment]
> Follow PEP8 standards in your code.
> Use spaces around operators and after commas, but not directly inside bracketing constructs.
> Read [this](https://www.geeksforgeeks.org/pep-8-coding-style-guide-python/)

@aashishpiitk You still haven't done this
